airsonic-main/src/main/java/org/airsonic/player/security/GlobalSecurityConfig.java

```diff
@@ -73,6 +78,32 @@ public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception
         auth.authenticationProvider(new JWTAuthenticationProvider(jwtKey));
     }
 
+    @Bean
+    public PasswordEncoder delegatingPasswordEncoder() {
+
+        // Spring Security 5 require storing the encoder id alongside the encoded password
+        // (e.g. "{md5}hash" for an MD5-encoded password hash), which differs from previous
+        // versions.
+        //
+        // Airsonic unfortunately stores passwords in plain-text, which is why we are setting
+        // the "no-op" (plain-text) password encoder as a default here. This default will be
+        // used when no encoder id is present.
+        //
+        // This means that legacy Airsonic passwords (stored simply as "password" in the db)
+        // will be matched like "{noop}password" and will be recognized successfully. In the
+        // future password encoding updates will be done here.
+
+        PasswordEncoder defaultEncoder = NoOpPasswordEncoder.getInstance();
+        String defaultIdForEncode = "noop";
+
+        Map<String, PasswordEncoder> encoders = new HashMap<>();
+        encoders.put(defaultIdForEncode, defaultEncoder);
+        DelegatingPasswordEncoder passworEncoder = new DelegatingPasswordEncoder(defaultIdForEncode, encoders);
+        passworEncoder.setDefaultPasswordEncoderForMatches(defaultEncoder);
+
+        return passworEncoder;
+    }
```

Comment:
This is upto you, but I think the easier and more correct way is to retain the default encoder and modify SecurityService to add "{noop}" to the password string retrieved from the db (see corresponding file in airsonic-advanced), so it's passed onto the correct delegate (which is already present in the default delegatingpasswordencoder).

Currently you're overriding the default password encoder that contains _all_ encoders (including noop) with something that only has a noop encoder.

you're effectively taking away all other password encoders, so the encoder is forced to use the only one there. instead i think you should take your implementation of passwords and route it to the correct place.

Additionally, please open an issue to migrate passwords to nonplaintext. I intend to submit a pr for that soon on airsonic-advanced.

Reply:
Your implementation, for me, has the issue that you're mixing up an implementation detail (adding "{noop}" to the password) in the middle of the authentication code, instead of doing it in the Spring configuration.

For the issue see #69 (you could have searched :wink:).

Reply:
It is absolutely an implementation detail of how passwords are implemented _specifically in airsonic_. Why would you therefore modify the framework and override the bean the framework provides?

Does the default spring password encoder not have the capability to decode the given password? Yes it does. And more. You just need to route it correctly.

The issue is specifically with airsonic implementation, therefore it seems to me that you need to modify the implementation itself to route the password correctly instead of change the framework to meet the incomplete implementation.

Consider future upgrades. Say while migrating you have a mix of bcrypt and scrypt and noop passwords coexisting in the db, which you identify in a separate column. Your configuration will be of no use. Rerouting the passwords to their correct encoder, however, will.

Thank you, that does seem to be an old issue. The password migration PR shouldn't be hard, but I will read the issue for potential pitfalls everyone seems to be trying to point out.

Reply:
Let me consider that some more, I'll have a closer look and see if there is a better way.

Reply:
I haven't yet looked at this in detail yet, but isnt this irrelevant given that we have a custom UserDetailsService which doesn't use a PasswordEncoder at all?

Reply:
No, if you use Spring's authentication (which you do), then you use its PasswordEncoder under the hood (which you do). The failing tests which @fxthomas no doubt observed alerted him to the issue leading to this change.

However, which PasswordEncoder you use is determined by a prefix to the password that, if we used Spring to encode prior to storage, is affixed to the password by Spring itself. You construct the user details with that password in your custom UserDetailsService (called SecurityService).

Because we don't use Spring to encode, and yet use it to authenticate and thus do password matching, Spring is at a loss at which encoder to forward the password to. It used to default to a noop strategy (thus is worked), but now the default is (I believe) bcrypt.

My proposal is, in lieu of migration, to affix the correct prefix that Spring itself would generate to the passwords. A bcrypt hash by Spring is stored as `{bcrypt}hash`, a noop would be stored as `{noop}password`; since we are storing only `password`, prefixing `{noop}` when feeding it to Spring (via UserDetails) lets it forward it to the correct encoder it has set up. The other encoders can also be redirected to with a different prefix. The other alternative implemented here is to change the Encoder bean itself and strip all encoders from it except the noop encoder (set it as default).

Reply:
I'm still considering what to do here. We also store MD5-encoded password like `enc:<md5 hash>` in that field (so it's not no-op at all in most cases), so neither solution looks good for me.

Essentially if we want to migrate to newer algorithms in the future it would make sense to create a `PasswordEncoder` that:

* If the encoded password starts with `{`, uses the default encoder
* If the encoded password starts with `enc:`, uses the MD5 encoder
* Otherwise uses the no-op decoder

Another question I still need to investigate is : do we really need the no-op decoder? I think we don't (it only seems to be used by the default admin user, changed at first boot), but I haven't completely verified that. If this is the case, it's even better to remove it entirely (which could be done in another PR).

Reply:
- I think you're mistaken: we do not store md5 hashes. That would render a particular form of auth mentioned in the topic you pointed to earlier (#69) inoperable (salt+token). We store the password itself in hex.
- The point with respect to additional "enc:" is well pointed, but, as currently constructed, it is opaque to the layer in question. "enc:" is just a DB persistence detail, completely abstracted away by the Dao; it isn't even seen by the auth code. The PasswordEncoder being used is absolutely the NoopEncoder, regardless of whatever and however you're storing things in the DB. You can of course make the Dao dumber and introduce a new Encoder that does what the Dao does, but after "decryption", it'll still be a NoopEncoder.
- Regardless of what you do or don't do in the DB, from my perspective, you need to migrate passwords into a consistent representation.

Reply:
> I think you're mistaken: we do not store md5 hashes

Indeed, you're right. I must have had a brain fart... Thanks for keeping up!

I still think the idea of having the correct Encoder is better in the sense that it is "correct" ; I'm uneasy about having "{noop}enc:something", this will bite us somewhere later. I'd rather get it right first.

Reply:
Again, I think you're losing something in abstraction, or I'm misunderstanding you.

> I'm uneasy about having "{noop}enc:something"

`{noop}enc:something` will never happen. `enc:` is a storage abstraction that only the dao is aware of. It never makes it out of the dao.

Maybe I can make it clearer:
- `plaintext` comes in
- `plaintext` goes to the dao to be stored
- dao stores it in the db as `enc:hex`, it really doesn't matter how the db stores it
- on retrieval (say for password checking), dao retrieves it and converts back to `plaintext`
- anything outside the dao (like the User object) never sees `enc:hex`, it only sees `plaintext`
- you now need to redirect `plaintext` to the correct password encoder for comparison, therefore you append `{noop}`

I hope that makes sense. How you store it (the domain of the dao) is different and distinct from how the password is compared (the domain of the password encoder). I could store it a million different ways and in a million different places, the comparison only takes whatever the User object has (which is always `plaintext`).

Reply:
Aaah, I think I understand what you mean now! I assumed (without checking, which I should have) that the User object had a 1:1 mapping with how it was stored in the DB, which is clearly not the case from what you say.